packages/cli/src/shutdown/Shutdown.service.ts

```diff
@@ -0,0 +1,85 @@
+import { Container, Service } from 'typedi';
+import { ApplicationError, ErrorReporterProxy, assert } from 'n8n-workflow';
+import { Logger } from '@/Logger';
+
+export interface ServiceClass {
+	new (): Record<string, () => Promise<void> | void>;
+}
+
+export interface ShutdownHandler {
+	serviceClass: ServiceClass;
+	methodName: string;
+}
+
+/** Error reported when a listener fails to shutdown gracefully */
+export class ComponentShutdownError extends ApplicationError {
+	constructor(componentName: string, cause: Error) {
+		super('Failed to shutdown gracefully', {
+			level: 'error',
+			cause,
+			extra: { component: componentName },
+		});
+	}
+}
+
+/** Service responsible for orchestrating a graceful shutdown of the application */
+@Service()
+export class ShutdownService {
+	readonly handlersByPriority: ShutdownHandler[][] = [];
```

Comment:
```suggestion
	private readonly handlersByPriority: ShutdownHandler[][] = [];
```

Reply:
We are checking this in the tests. I don't personally like having fields public just for the sake of testing. I changed it private and added an `@ts-expect-error` to the test. I think it's better that way.